File: ios/MidTrim/MidTrim/Domain/UseCases/TrimVideoUseCase.swift
import Foundation

struct TrimVideoUseCase {
    private let trimmer: VideoTrimmerProtocol
    private let trimValidator: ValidateTrimDurationUseCase
    private let windowCalculator: CalculateTrimWindowUseCase

    init(trimmer: VideoTrimmerProtocol, trimValidator: ValidateTrimDurationUseCase, windowCalculator: CalculateTrimWindowUseCase) {
        self.trimmer = trimmer
        self.trimValidator = trimValidator
        self.windowCalculator = windowCalculator
    }

    func execute(sourceURI: String, duration: Double, trimDuration: Double, isPaidUser: Bool) async throws -> String {
        guard !sourceURI.isEmpty else {
            throw TrimError.fileNotFound("")
        }
        try trimValidator.execute(trimDuration: trimDuration, isPaidUser: isPaidUser)

        let window = try windowCalculator.execute(videoDuration: duration, trimDuration: trimDuration)
        return try await trimmer.trim(sourceURI: sourceURI, startTime: window.startTime, endTime: window.endTime)
    }
}

File: android/app/src/main/java/com/izzatismail/midtrim/domain/usecase/TrimVideoUseCase.kt
package com.izzatismail.midtrim.domain.usecase

import com.izzatismail.midtrim.domain.error.TrimError
import com.izzatismail.midtrim.domain.repository.VideoTrimmer

class TrimVideoUseCase(
    private val trimmer: VideoTrimmer,
    private val trimValidator: ValidateTrimDurationUseCase,
    private val windowCalculator: CalculateTrimWindowUseCase
) {
    suspend fun execute(
        sourceUri: String,
        videoDuration: Double,
        trimDuration: Double,
        isPaidUser: Boolean
    ): String {
        require(sourceUri.isNotEmpty()) { "sourceUri must not be empty" }
        trimValidator.execute(trimDuration, isPaidUser)
        val window = windowCalculator.execute(videoDuration, trimDuration)
        return trimmer.trim(sourceUri, window.startTime, window.endTime)
    }
}